docs/design/protobuf-grpc-service.md

```diff
@@ -0,0 +1,162 @@
+
+- Some of the Kubernetes API like `tolerance` and `node affinity` are too complicated to be converted to an API.
+- We want to leave some flexibility to use database to store history data in the near future (for example, pagination, list options etc).
+
+We end up propsing a simple and easy API which can cover most of the daily requirements.
+
+```
+message Cluster {
+  // Required input field. Unique Cluster name provided by user.
+  string name = 1;
+
+  // Required input field. Cluster's namespace provided by user
+  string namespace = 2;
+
+  // Required field. This field indicates the user who owns the cluster.
+  string user = 3;
+
+  // Optional input field. Ray cluster version
+  string version = 4;
+
+  // Optional field.
+  enum Environment {
+    DEV = 0;
+    TESTING = 1;
+    STAGING = 2;
+    PRODUCTION = 3;
+  }
+  Environment environment = 5;
+
+  // Required field. This field indicates ray cluster configuration
+  ClusterSpec cluster_spec = 6;
+
+  // Output. The time that the Cluster created.
+  google.protobuf.Timestamp created_at = 7;
+
+  // Output. The time that the Cluster deleted.
+  google.protobuf.Timestamp deleted_at = 8;
+}
+
+message ClusterSpec {
+  // The head group configuration
+  HeadGroupSpec head_group_spec = 1;
+  // The worker group configurations
+  repeated WorkerGroupSpec worker_group_sepc = 2;
+}
+
+message HeadGroupSpec {
+  // Optional. The computeTemplate of head node group
+  string compute_template = 1;
+  // Optional field. This field will be used to retrieve right ray container
+  string image = 2;
+  // Optional. The service type (ClusterIP, NodePort, Load balancer) of the head node
+  string service_type = 3;
+  // Optional. The ray start parames of head node group
+  map<string, string> ray_start_params = 4;
+}
+
+message WorkerGroupSpec {
+  // Required. Group name of the current worker group
+  string group_name = 1;
+  // Optional. The computeTemplate of head node group
+  string compute_template = 2;
+  // Optional field. This field will be used to retrieve right ray container
+  string image = 3;
+  // Required. Desired replicas of the worker group 
+  int32 replicas = 4;
+  // Optional. Min replicas of the worker group 
+  int32 min_replicas = 5;
+  // Optional. Max replicas of the worker group 
+  int32 max_replicas = 6;
+  // Optional. The ray start parames of worker node group
+  map<string, string> ray_start_params = 7;
+}
+
+// ComputeTemplate can be reused by any compute units like worker group, workspace, image build job, etc
+message ComputeTemplate {
+  // The ID of the compute template
+  string name = 1;
+  // Number of cpus
+  uint32 cpu = 2;
+  // Number of memory
+  uint32 memory = 3;
+  // Number of gpus
+  uint32 gpu = 4;
+  // The detail gpu accelerator type
+  string gpu_accelerator = 5;
+}
+
+// ImageTemplate can be used by worker group and workspce.
+// They can be distinguish by different entrypoints
+message ImageTemplate {
```

Comment:
I think ray-operator should not be aware of the imageTemplate. it just need `image` to build the cluster.  `ImageTemplate` corresponding services can be used to build images separately. 


1. User who does not have docker experience can build images using this service. 
2. User can query the image result built by the service
3. User can use that image result to build ray cluster.